core/primitives/transaction_validity.hpp

```diff
@@ -53,123 +54,139 @@ namespace kagome::primitives {
-    }
-  };
-
-  /// Transaction validity can't be determined.
-  struct Unknown {
-    uint8_t error;
-
-    bool operator==(const Unknown &rhs) const {
-      return error == rhs.error;
-    }
-    bool operator!=(const Unknown &rhs) const {
-      return error != rhs.error;
-    }
+  enum class InvalidTransaction : uint8_t {
+    /// The call of the transaction is not expected.
+    Call,
+    /// General error to do with the inability to pay some fees (e.g. account
+    /// balance too low).
+    Payment,
+    /// General error to do with the transaction not yet being valid (e.g. nonce
+    /// too high).
+    Future,
+    /// General error to do with the transaction being outdated (e.g. nonce too
+    /// low).
+    Stale,
+    /// General error to do with the transaction's proofs (e.g. signature).
+    BadProof,
+    /// The transaction birth block is ancient.
+    AncientBirthBlock,
+    /// The transaction would exhaust the resources of current block.
+    ///
+    /// The transaction might be valid, but there are not enough resources left
+    /// in the current block.
+    ExhaustsResources,
+    /// Any other custom invalid validity that is not covered by this enum.
+    Custom
   };
 
-  /**
-   * Information on a transaction's validity and, if valid, on how it relate to
-   * other transactions.
-   */
-  using TransactionValidity = boost::variant<Invalid, Valid, Unknown>;
-
-  /**
-   * @brief outputs object of type Invalid to stream
-   * @tparam Stream stream type
-   * @param s stream reference
-   * @param v value to output
-   * @return reference to stream
-   */
   template <class Stream,
             typename = std::enable_if_t<Stream::is_encoder_stream>>
-  Stream &operator<<(Stream &s, const Invalid &v) {
-    return s << v.error;
+  Stream &operator<<(Stream &s, const InvalidTransaction &v) {
+    return s << static_cast<uint8_t>(v);
   }
 
-  /**
-   * @brief outputs object of type Valid to stream
-   * @tparam Stream stream type
-   * @param s stream reference
-   * @param v value to output
-   * @return reference to stream
-   */
   template <class Stream,
-            typename = std::enable_if_t<Stream::is_encoder_stream>>
-  Stream &operator<<(Stream &s, const Valid &v) {
-    return s << v.priority << v.requires << v.provides << v.longevity;
+            typename = std::enable_if_t<Stream::is_decoder_stream>>
+  Stream &operator>>(Stream &s, InvalidTransaction &v) {
+    uint8_t value = 0u;
+    s >> value;
+    if (value > static_cast<uint8_t>(InvalidTransaction::ExhaustsResources)) {
+      v = InvalidTransaction::Custom;
+    } else {
+      v = static_cast<InvalidTransaction>(value);
+    }
+    return s;
   }
 
-  /**
-   * @brief outputs object of type Unknown to stream
-   * @tparam Stream stream type
-   * @param s stream reference
-   * @param v value to output
-   * @return reference to stream
-   */
+  /// An unknown transaction validity.
+  enum class UnknownTransaction : uint8_t {
```

Comment:
Values should start with 1 and also naming is unconventional.

Reply:
same